apps/system/utils/kdbg_heapinfo.c

```diff
@@ -114,68 +279,91 @@ int kdbg_heapinfo(int argc, char **args)
 {
 #ifdef CONFIG_DEBUG_MM_HEAPINFO
 	int option;
-	int display_flag = HEAPINFO_DISPLAY_ALL;
-	int mode = HEAPINFO_SIMPLE;
-	int pid = HEAPINFO_PID_ALL;
-#if CONFIG_MM_REGIONS > 1
-	int region_idx;
+#ifdef CONFIG_MM_KERNEL_HEAP
+	bool showing_kheap = false;
+	heapinfo_option_t kernel_option;
 #endif
 #if CONFIG_MM_NHEAPS > 1
 	bool summary_option = true;
 	int heap_idx = -1;
+	int region_num;
 #endif
-
 	if (argc >= 2 && !strncmp(args[1], "--help", strlen("--help") + 1)) {
 		goto usage;
 	}
 
+	heapinfo_display_flag = HEAPINFO_DISPLAY_ALL;
+	heapinfo_mode = HEAPINFO_SIMPLE;
+	heapinfo_pid = HEAPINFO_PID_ALL;
+
 #if CONFIG_MM_NHEAPS > 1
-	kdbg_init_total_info();
+	heapinfo_init_totalinfo();
 #endif
-	struct mm_heap_s *user_heap = g_mmheap;
-	while ((option = getopt(argc, args, "iap:fge:r")) != ERROR) {
+	struct mm_heap_s *heap = g_mmheap;
+	while ((option = getopt(argc, args, "iap:fge:rk:")) != ERROR) {
 #if CONFIG_MM_NHEAPS > 1
 		summary_option = false;
 #endif
 		switch (option) {
 		case 'i':
-			sched_foreach(kdbg_heapinfo_init, NULL);
+			heapinfo_peak_init(heap);
 			printf("Peak allocated memory size is cleared\n");
 			return OK;
 		case 'a':
-			mode = HEAPINFO_DETAIL_ALL;
-			pid = HEAPINFO_PID_ALL;
-			display_flag = HEAPINFO_DISPLAY_ALL;
+			heapinfo_set_option('a', HEAPINFO_PID_ALL);
 #if CONFIG_MM_NHEAPS > 1
 			summary_option = true;
 #endif
 			break;
 		case 'p':
-			mode = HEAPINFO_DETAIL_PID;
-			pid = atoi(optarg);
-			display_flag = HEAPINFO_DISPLAY_ALL;
+			if (atoi(optarg) < 0) {
+				printf("Invalid PID.\n");
+				goto usage;
+			}
+			heapinfo_set_option('p', atoi(optarg));
 			break;
 		case 'f':
-			mode = HEAPINFO_DETAIL_FREE;
-			pid = HEAPINFO_PID_ALL;
-			display_flag = HEAPINFO_DISPLAY_ALL;
+			heapinfo_set_option('f', HEAPINFO_PID_ALL);
 			break;
 		case 'g':
-			display_flag = HEAPINFO_DISPLAY_GROUP;
+			heapinfo_set_option('g', HEAPINFO_PID_ALL);
 			break;
 		case 'e':
 #if CONFIG_MM_NHEAPS > 1
-			mode = HEAPINFO_DETAIL_SPECIFIC_HEAP;
+			heapinfo_set_option('e', HEAPINFO_PID_ALL);
 			heap_idx = atoi(optarg);
-			display_flag = HEAPINFO_DISPLAY_SPECIFIC_HEAP;
 #else
 			goto usage;
 #endif
 			break;
 		case 'r':
 #if CONFIG_MM_REGIONS > 1
-			display_flag = HEAPINFO_DISPLAY_REGION;
+			heapinfo_print_regions();
+#else
+			goto usage;
+#endif
+			break;
+		case 'k':
+#ifdef CONFIG_MM_KERNEL_HEAP
+			heapinfo_fd = open(HEAPINFO_DRVPATH, O_RDWR);
+			if (heapinfo_fd < 0) {
+				printf("Fail to open heapinfo driver %d.\n", errno);
+				return ERROR;
+			}
+			showing_kheap = true;
+			kernel_option.mode = heapinfo_mode; 
+			kernel_option.pid = heapinfo_pid;
+			if (strncmp(optarg, "0", strlen("0")) == 0 || atoi(optarg)) {
+				/* If option is digit, show heapinfo per pid in kernel heap */
+				heapinfo_set_option('p', atoi(args[2]));
+			} else if (strncmp(optarg, "i", strlen("i")) == 0) {
```

Comment:
If optarg includes kinds of '-i', '-a', something wrong would happen.

Reply:
If optarg includes kinds of '-i', '-a', it will be entered to default option.(It will print "Invalid option")

Reply:
I changed not to handle the optarg.